sympy/matrices/expressions/determinant.py

```diff
@@ -40,11 +40,19 @@ def arg(self):
     def kind(self):
         return self.arg.kind.element_kind
 
-    def doit(self, expand=False, **hints):
-        try:
-            return self.arg._eval_determinant()
-        except (AttributeError, NotImplementedError):
-            return self
+    def doit(self, **hints):
+        arg = self.arg
+        if hints.get('deep', True):
+            arg = arg.doit(**hints)
+
+        _eval_determinant = getattr(arg, '_eval_determinant', None)
```

Comment:
I try to get an review of this, but usually, `hasattr` should be avoided as well as the `AttributeError` catch.

Reply:
I prefer `getattr(obj, attrname, None)` to either `hasattr` or catching `AttributeError`.

It is a bad sign that the existence of the method needs to be tested at all though.

Reply:
I'd anyway review about the general problem of `doit` though. I don't think that it encourages applying `deep` consistently.

Reply:
Thanks for the reviews.

This would be an alternative to the dynamic access with `getattr`/`hasattr` or catching `AttributeError`:

```python
if isinstance(arg, MatrixExpr):
      result = arg._eval_determinant()
```

`_eval_determinant` is defined in `MatrixExpr` and as far as I know, all classes which implement `_eval_determinant` inherit from `MatrixExpr`. In my opinion the check for the instance is the best option.

@sylee957 I'm not sure what you mean by reviewing about the *general* problem of `doit`. Do you want to rewrite all `doit` methods in a consistent way? If so, this pull request might be too specific and I would close it.

Reply:
The Determinant constructor sympifies its argument and then checks for a matrix. Under what situation would the argument to Determinant not have an `_eval_determinant` method?

Reply:
`sympy.matrices.common._MinimalMatrix` is a Matrix (`is_Matrix = True`), but it does not implement `_eval_determinant`. However, `_MinimalMatrix` is not public and is only used for testing.

Every matrix the user is working with, implements `_eval_determinant`. I think it would be okay, to skip the check if `_eval_determinant` exists.

Reply:
Furthermore, sympifying a `_MinimalMatrix` throws an exception anyway. Consequently, an exception is already raised in the constructor of `Determinant`